### PR TITLE
Make some basic THC operations thread-safe

### PR DESCRIPTION
**Pull request by colesbury**
Switching the device, setting the stream, and switching BLAS handles is
now thread-safe. Some other operations, like reserveStreams, are still
not thread-safe.

### CODE REVIEW

File: init.c [hunk 1]
```diff
@@ -955,18 +948,17 @@ int luaopen_libcutorch(lua_State *L)
   lua_setglobal(L, "cutorch");
   luaL_setfuncs(L, cutorch_stuff__, 0);
 
-  THCState* state = (THCState*)malloc(sizeof(THCState));
-  memset(state, 0, sizeof(THCState));
+  THCState* state = THCState_alloc();
```
Comment (apaszke) [inline init.c]: `THCState_new` would be more consistent with existing APIs.

File: lib/THC/THCGeneral.c [hunk 2]
```diff
@@ -3,56 +3,121 @@
...
+  THCCudaResourcesPerDevice* resourcesPerDevice;
+  /* Captured number of devices upon startup; convenience for bounds checking */
+  int numDevices;
+  /* Number of Torch defined resources available, indices 1 ... numStreams */
+  int numUserStreams;
+  int numUserBlasHandles;
+
+  /* Allocator using cudaMallocHost. */
+  THAllocator* cudaHostAllocator;
+  THCDeviceAllocator cudaDeviceAllocator;
```
Comment (apaszke) [inline lib/THC/THCGeneral.c]: Why is the device allocator a member, while the host allocator is referenced via a pointer? Wouldn't it be more consistent to have both of them point to static structs with function pointers (like in TH)?

File: lib/THC/THCGeneral.c [hunk 3]
```diff
@@ -3,56 +3,121 @@
...
   state->deviceProperties =
-    (struct cudaDeviceProp*)malloc(count * sizeof(struct cudaDeviceProp));
+    (struct cudaDeviceProp*)malloc(numDevices * sizeof(struct cudaDeviceProp));
 
-  state->numUserStreams = 0;
-  state->numUserBlasHandles = 0;
+  state->rngState = (THCRNGState*)malloc(sizeof(THCRNGState));
+  THCRandom_init(state, numDevices, device);
+
+  state->cudaHostAllocator = (THAllocator*)malloc(sizeof(THAllocator));
```
Comment (apaszke) [inline lib/THC/THCGeneral.c]: This can be a static struct.

File: lib/THC/THCThreadLocal.c [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+void THCThreadLocal_free(THCThreadLocal local)
+{
+#ifndef _WIN32
+  THAssert(pthread_key_delete(local) == 0);
+#else
+  THAssert(TlsFree(local));
+#endif
+}
+
+void* THCThreadLocal_get(THCThreadLocal local)
```
Comment (apaszke) [inline lib/THC/THCThreadLocal.c]: This is an internal class, and we're using C++, so can we just use the templates to determine the held type. I understand that we might save on one or two malloc/free calls, but I'm not sure if it's worth complicating the code. Let's just store an `int*` pointing to a single value and return it a properly typed pointer/reference from get. If we really want to avoid the malloc, we can just specialize this class for small types (and hide all intptr_t casts inside this class).

File: lib/THC/THCGeneral.h.in [hunk 5]
```diff
@@ -57,50 +47,10 @@ typedef struct _THCDeviceAllocator {
...
-  int p2pKernelAccessEnabled;
-
-  void (*cutorchGCFunction)(void *data);
-  void *cutorchGCData;
-  long heapSoftmax;
-  long heapDelta;
-} THCState;
+typedef struct THCState THCState;
+
+THC_API THCState* THCState_alloc();
```
Comment (apaszke) [inline lib/THC/THCGeneral.h.in]: `new`?

File: lib/THC/THCThreadLocal.h [hunk 6]
```diff
@@ -0,0 +1,16 @@
+#ifndef THC_THREAD_LOCAL_INC
+#define THC_THREAD_LOCAL_INC
+
+#ifdef _WIN32
+typedef DWORD THCThreadLocal;
+#else
+#include <pthread.h>
```
Comment (apaszke) [inline lib/THC/THCThreadLocal.h]: It might be better to add something like [this](https://github.com/torch/torch7/blob/master/lib/TH/CMakeLists.txt#L188-L193) to CMakeLists.txt if using pthread.

File: lib/THC/THCGeneral.c (hunk 3 above)
Comment (colesbury) [inline lib/THC/THCGeneral.c]: Then you would be calling `THCAllocator_init()` on a static (global) variable. All of THC is per-THCState

Comment (colesbury) [inline lib/THC/THCGeneral.c]: (you could make it a static struct that lives in THCAllocator.c, but this is all unrelated changes to THC. The short answer is that this is how it was before this PR)

File: lib/THC/THCThreadLocal.c (hunk 4 above)
Comment (colesbury) [inline lib/THC/THCThreadLocal.c]: We're not using C++ here or in THCGeneral.c. Adding templates doesn't seem like it'll **reduce** code complication, especially since you can't really specialize the POSIX or Windows thread-local API. They both take `void*`, not arbitrary types.

File: init.c (hunk 1 above)
Comment (colesbury) [inline init.c]: It's `alloc` because there's a separate initialization function. The `_new` functions initialize their data structure.
